AdyenCardScanner/Sources/CardImageParser.swift

```diff
@@ -90,84 +90,36 @@ class CardImageParser: CardImageParsing {
-        let colorControlsFilter = CIFilter(name: "CIColorControls")
-        colorControlsFilter?.setValue(image, forKey: kCIInputImageKey)
-        colorControlsFilter?.setValue(0.0, forKey: kCIInputSaturationKey) // No color
-        colorControlsFilter?.setValue(1.0, forKey: kCIInputBrightnessKey) // Full brightness
-        colorControlsFilter?.setValue(1.0, forKey: kCIInputContrastKey) // Full contrast
-        return colorControlsFilter?.outputImage
-    }
-
-    private func applyThreshold(image: CIImage, threshold: CGFloat) -> CIImage? {
-        let colorMatrixFilter = CIFilter(name: "CIColorMatrix")
-        colorMatrixFilter?.setValue(image, forKey: kCIInputImageKey)
-
-        // Set the threshold value
-        let thresholdVector = CIVector(x: threshold, y: threshold, z: threshold, w: 1)
-        colorMatrixFilter?.setValue(thresholdVector, forKey: "inputRVector")
-        colorMatrixFilter?.setValue(thresholdVector, forKey: "inputGVector")
-        colorMatrixFilter?.setValue(thresholdVector, forKey: "inputBVector")
-
-        // Keeping alpha channel as is
-        colorMatrixFilter?.setValue(CIVector(x: 0, y: 0, z: 0, w: 1), forKey: "inputAVector")
-
-        // Biasing the output so that it’s either black or white
-        let biasVector = CIVector(x: -threshold, y: -threshold, z: -threshold, w: 0)
-        colorMatrixFilter?.setValue(biasVector, forKey: "inputBiasVector")
-
-        return colorMatrixFilter?.outputImage
-    }
-
-    // ==================================================
-
     private func extractCardNumber(from textObservations: [VNRecognizedTextObservation]) -> String? {
-        for observation in textObservations {
-            let candidates = observation.topCandidates(Constants.topCandidates)
-            guard let topCandidate = candidates.first,
-                  topCandidate.confidence > Constants.cardNumberConfidence else { continue }
-
-            debugPrint("candidates >> \(topCandidate.string)")
-
-            let sanitizedCardNumber = topCandidate.string.replacingOccurrences(of: " ", with: "")
-
-            guard sanitizedCardNumber.isOnlyNumbers else { continue }
-            guard sanitizedCardNumber.count >= 13, sanitizedCardNumber.count <= 19 else { continue }
-            guard isValidLuhn(sanitizedCardNumber) else { continue }
-
-            debugPrint("VALID CARD: \(sanitizedCardNumber)")
-
-            return sanitizedCardNumber
-        }
-
-        return nil
+        if let cachedCardNumber { return cachedCardNumber }
+
+        let cardNumberMatch = textObservations
+            .compactMap { $0.topCandidates(Constants.topCandidates).first }
+            .filter { $0.confidence > Constants.cardNumberConfidence }
+            .map { $0.string.replacingOccurrences(of: " ", with: "") }
+            .filter { $0.isOnlyNumbers }
+            .filter { $0.count >= 13 && $0.count <= 19 }
+            .filter { isValidLuhn($0) }
```

Comment:
Maybe convert this 3 into a single check?

```swift
.filter { $0.isCardNumber }

...

extension String {
    internal var isCardNumber: Bool {
        self.isOnlyNumbers && (13...19).contains(self.count) &&  self.isValidLuhn
    }
    private var isOnlyNumbers: Bool { ... }
    private var isValidLuhn: Bool { ... }
}
```

Reply:
I can do, good suggestion